File: backend/app/models/telemetry.py
from datetime import datetime

from pydantic import BaseModel, Field


class Location(BaseModel):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    altitude_meters: float | None = None
    accuracy_meters: float = Field(ge=0)


class Accelerometer(BaseModel):
    x: float
    y: float
    z: float


class Gyroscope(BaseModel):
    x: float
    y: float
    z: float


class TelemetryPayload(BaseModel):
    version: str
    device_id: str
    timestamp: datetime
    location: Location
    speed_mps: float | None = Field(default=None, ge=0)
    bearing_degrees: float | None = Field(default=None, ge=0, lt=360)
    accelerometer: Accelerometer
    gyroscope: Gyroscope
    battery_level: float = Field(ge=0.0, le=1.0)
    network_status: str

File: mobile/app/src/main/java/com/wegotchu/app/data/model/Telemetry.kt
package com.wegotchu.app.data.model

import com.google.gson.annotations.SerializedName

data class Location(

    val latitude: Double,

    val longitude: Double,

    @SerializedName("altitude_meters")
    val altitudeMeters: Double?,

    @SerializedName("accuracy_meters")
    val accuracyMeters: Double
)

data class Accelerometer(

    val x: Double,

    val y: Double,

    val z: Double
)

data class Gyroscope(

    val x: Double,

    val y: Double,

    val z: Double
)

data class TelemetryPayload(

    val version: String,

    @SerializedName("device_id")
    val deviceId: String,

    val timestamp: String,

    val location: Location,

    @SerializedName("speed_mps")
    val speedMps: Double?,

    @SerializedName("bearing_degrees")
    val bearingDegrees: Double?,

    val accelerometer: Accelerometer,

    val gyroscope: Gyroscope,

    @SerializedName("battery_level")
    val batteryLevel: Double,

    @SerializedName("network_status")
    val networkStatus: String
)